### PR TITLE
Improved SafeERC20 allowance handling

### LINKED ISSUE #1404: Improve SafeERC20 allowance handling
Currently, `SafeERC20`'s allowance-related functionality is restricted to a `safeApprove` function, which is vulnerable to the well-known non-zero approval issue in the ERC20 spec. We should:

- add `safeIncreaseAllowance` and `safeDecreaseAllowance`, with `require` on the return value of the `ERC20` functions
- add extra `require`s to `safeApprove` so that it only works if the new or old allowance are 0 (i.e. it only sets an initial allowance, or sets it to 0, which are the only safe scenarios).

This way, we'll be guiding users to the best practices regarding `ERC20` tokens.

Props to @cwhinfrey and the Level K team for suggesting this!

### PR DESCRIPTION
**Pull request by Aniket-Engg**
Fixes #1404

### CODE REVIEW

File: contracts/token/ERC20/IERC20.sol [hunk 1]
```diff
@@ -20,6 +20,12 @@ interface IERC20 {
   function transferFrom(address from, address to, uint256 value)
     external returns (bool);
 
+  function increaseAllowance(address spender, uint256 addedValue)
```
Comment (nventuro) [inline contracts/token/ERC20/IERC20.sol]: We shouldn't add these here, since this is just the ERC20 specification, which doesn't include these methods. I'm actually unsure about this, maybe we should have an additional interface that describes these? (e.g. `IERC20Allowance`). Let's leave it like this for now, though.

File: contracts/token/ERC20/SafeERC20.sol [hunk 2]
```diff
@@ -38,6 +38,27 @@ library SafeERC20 {
   )
     internal
   {
+    require((value == 0) || (token.allowance(msg.sender, spender) == 0));
```
Comment (nventuro) [inline contracts/token/ERC20/SafeERC20.sol]: We may want to add a comment here, saying something like:

'safeApprove should only be called when setting an initial allowance, or when resetting it to zero. To increase and decrease it, use safeIncreaseAllowance and safeDecreaseAllowance.'

File: contracts/mocks/SafeERC20Helper.sol [hunk 3]
```diff
@@ -46,12 +54,20 @@ contract ERC20SucceedingMock is IERC20 {
...
+    return true;
+  }
+
   function balanceOf(address) public view returns (uint256) {
     return 0;
   }
 
   function allowance(address, address) public view returns (uint256) {
-    return 0;
+    return 10;  //non-zero allowance
```
Comment (nventuro) [inline contracts/mocks/SafeERC20Helper.sol]: What if we instead add a public function to set this value? That way, we can test that `approve` fails whenever one of the old and new allowances aren't zero.

File: contracts/token/ERC20/IERC20.sol (hunk 1 above)
Comment (Aniket-Engg) [reply contracts/token/ERC20/IERC20.sol]: I vote for additional interface named as `ISafeERC20` or `IERC20Allowance`

File: contracts/token/ERC20/SafeERC20.sol (hunk 2 above)
Comment (Aniket-Engg) [reply contracts/token/ERC20/SafeERC20.sol]: will add that.

File: contracts/mocks/SafeERC20Helper.sol (hunk 3 above)
Comment (Aniket-Engg) [reply contracts/mocks/SafeERC20Helper.sol]: Usually `approve` does that. But here we have a bit different case. So we can add a `setAllowance` function which will set a private variable `allowance` and this variable will be returned by function `allowance`. What you think?

Comment (nventuro) [reply contracts/mocks/SafeERC20Helper.sol]: Yes, that sounds great! :)

File: contracts/token/ERC20/IERC20.sol (hunk 1 above)
Comment (Aniket-Engg) [reply contracts/token/ERC20/IERC20.sol]: done

File: contracts/mocks/SafeERC20Helper.sol (hunk 3 above)
Comment (Aniket-Engg) [reply contracts/mocks/SafeERC20Helper.sol]: done

**Comment by Aniket-Engg** [comment]
I think, build needs to re-run.

File: contracts/token/ERC20/IERC20.sol (hunk 1 above)
Comment (nventuro) [reply contracts/token/ERC20/IERC20.sol]: @Aniket-Engg an even better solution would be to implement the methods inside the library, since that would allow using them with non-OpenZeppelin tokens. Take a look at [this snippet](https://gist.github.com/cwhinfrey/91f63c4f0471308f63d7657f8339062e#file-safeerc20-sol-L49) for how to do so. Thanks @cwhinfrey for the suggestion!

File: contracts/token/ERC20/IERC20.sol [hunk 4]
```diff
@@ -20,6 +20,15 @@ interface IERC20 {
   function transferFrom(address from, address to, uint256 value)
     external returns (bool);
 
+  function increaseAllowance(address spender, uint256 addedValue) 
+    external returns (bool);
+
+  function decreaseAllowance(address spender, uint256 subtractedValue) 
+    external returns (bool);
+
+  function setAllowance(uint256 value)
```
Comment (nventuro) [inline contracts/token/ERC20/IERC20.sol]: The tests are failing because this function is making `ERC20` abstract (since the method is not implemented). This should only be added to the mock contracts, not the standard interface.

Comment (Aniket-Engg) [reply contracts/token/ERC20/IERC20.sol]: got it.

File: contracts/token/ERC20/IERC20.sol (hunk 1 above)
Comment (Aniket-Engg) [reply contracts/token/ERC20/IERC20.sol]: @nventuro i did that but we still need a better way set non-zero allowance for an address. I tried something but they were not compatible with standard and consistency so i eliminated them. pushed remaining code.

**Comment by nventuro** [comment]
@Aniket-Engg thank you for your great help, as always! We need to get this fix in soon since we plan on finishing the release tomorrow, so I pushed directly to your branch - sorry about that!

**Comment by Aniket-Engg** [comment]
No problem at all @nventuro
